RCP checker refactor: reading submission

maanug-nv opened the pull request:
Refactor `get_submission_epochs()`:
 - function now infers benchmark from results files instead of taking as a parameter (resolves #254)
 - checks for "eval_error" string directly in logs since resulting behavior is the same (no need to call `_detect_eval_error()`
 - Ensure batch size and benchmark name are consistent across results files

Depends #289

pgmpablo157321 commented:
@maanug-nv I get this error when running the package checker: 
```
INFO - ------------------------------
INFO -  Running RCP Checker, pass: pruned_rcps
INFO - ------------------------------
Traceback (most recent call last):
  File "/Users/pablogonzalez/opt/anaconda3/envs/checker/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/pablogonzalez/opt/anaconda3/envs/checker/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/pablogonzalez/Desktop/MLCommons/mlperf-logging/mlperf_logging/package_checker/__main__.py", line 3, in <module>
    package_checker.main()
  File "/Users/pablogonzalez/Desktop/MLCommons/mlperf-logging/mlperf_logging/package_checker/package_checker.py", line 286, in main
    check_training_package(args.folder, args.usage, args.ruleset, args.quiet, args.werror,
  File "/Users/pablogonzalez/Desktop/MLCommons/mlperf-logging/mlperf_logging/package_checker/package_checker.py", line 217, in check_training_package
    check_training_result_files(folder, usage, ruleset, quiet, werror, rcp_bypass, rcp_bert_train_samples)
  File "/Users/pablogonzalez/Desktop/MLCommons/mlperf-logging/mlperf_logging/package_checker/package_checker.py", line 168, in check_training_result_files
    rcp_pass, rcp_msg = rcp_chk.check_directory(benchmark_folder, rcp_pass='pruned_rcps', rcp_bypass=rcp_bypass, set_scaling=True)
  File "/Users/pablogonzalez/Desktop/MLCommons/mlperf-logging/mlperf_logging/rcp_checker/rcp_checker.py", line 458, in check_directory
    rcp_check = self._eval_submission_record(interp_record, subm_epochs, (dir if set_scaling else ''))
  File "/Users/pablogonzalez/Desktop/MLCommons/mlperf-logging/mlperf_logging/rcp_checker/rcp_checker.py", line 398, in _eval_submission_record
    subm_epochs.sort()
TypeError: '<' not supported between instances of 'NoneType' and 'float'
```
Logging some extra values you can see that this variable can have some `None` values:
```
INFO - FILES: ['saved_files/submission-ui-tests/training/asus/ASUSTeK/results/ESCN4A-E11/dlrm/result_4.txt', 'saved_files/submission-ui-tests/training/asus/ASUSTeK/results/ESCN4A-E11/dlrm/result_0.txt', 'saved_files/submission-ui-tests/training/asus/ASUSTeK/results/ESCN4A-E11/dlrm/result_1.txt', 'saved_files/submission-ui-tests/training/asus/ASUSTeK/results/ESCN4A-E11/dlrm/result_3.txt', 'saved_files/submission-ui-tests/training/asus/ASUSTeK/results/ESCN4A-E11/dlrm/result_2.txt']
INFO - SUBM_EPOCHS: [1.9, 1.9, 1.9, 1.9, None]
```
This happens because one of the results may have errors and still be (as I understand) a valid submission.

maanug-nv commented:
I'll try to repro. Is this test from 2.1 submissions?

maanug-nv commented:
I was able to repro, thanks for finding this. I'll work on a fix, I think it is from the initial value of `subm_epochs = None` in `read_submission_file()`. Will need to determine the right default.

xyhuang commented on mlperf_logging/rcp_checker/rcp_checker.py:
```diff
@@ -37,64 +38,83 @@
+                logging.warning(' Batch sizes in files do not match.')
+                return -1, None, None
+
+        if benchmark is None:
```
it seems benchmark is always going to be None here since it is initialized to None in line 92, is this intended? is the benchmark variable still useful here?

xyhuang commented on mlperf_logging/rcp_checker/rcp_checker.py:
```diff
@@ -490,5 +509,29 @@ def make_checker(usage, ruleset, verbose=False, bert_train_samples=False):
+    # Check pruned RCPs by default. Use rcp_pass='full_rcp' for full check
+    test, msg = checker.check_directory(args.dir, rcp_pass=args.rcp_pass)
+
+    if test:
```
minor nit: maybe name it "passed" or something like that to be more intuitive

maanug-nv commented:
I think the resolution to the issue Pablo mentioned is to mark the result as infinity/`1e9` which will cause the olympic scoring (done later in the checker after `get_submission_epochs()`) to drop this result. I don't think it should be dropped in this function.

pgmpablo157321 commented:
> I think the resolution to the issue Pablo mentioned is to mark the result as infinity/1e9 which will cause the olympic scoring (done later in the checker after get_submission_epochs()) to drop this result. I don't think it should be dropped in this function.

Ok, that sounds reasonable

maanug-nv commented:
@pgmpablo157321 I've added a warning and changed the defaults of `not_converged` and `subm_epochs` to behave as I previously described. I also reran the package checker on a few submissions, it looks good to me now.

maanug-nv commented:
Is there anything blocking this from merge? This needs to be merged first before #294
